### PR TITLE
[runtime] Convert ValueType icalls to handles.

### PR DESCRIPTION
**Pull request by vargaz**

### CODE REVIEW

**Comment by vargaz** [comment]
@monojenkins build failed

File: mono/metadata/icall.c [hunk 1]
```diff
@@ -7602,20 +7585,23 @@ mono_icall_get_logical_drives (void)
...
 	ndrives = 0;
 	do {
 		len = 0;
 		u16 = dname;
-		while (*u16) { u16++; len ++; }
-		drivestr = mono_string_new_utf16_checked (domain, dname, len, error);
-		if (mono_error_set_pending_exception (error))
-			goto leave;
+		while (*u16) {
+			u16++; len ++;
```
Comment (jaykrell) [inline mono/metadata/icall.c]: `g_utf16_len`

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
Helix problems?

**Comment by CoffeeFlux** [comment]
@monojenkins build failed

File: mono/metadata/icall.c [hunk 2]
```diff
@@ -1347,69 +1346,64 @@ ves_icall_System_ValueType_InternalGetHashCode (MonoObject *this_obj, MonoArray
...
 		default:
 			if (!values)
-				values = g_newa (MonoObject*, mono_class_num_fields (klass));
-			o = mono_field_get_value_object_checked (mono_object_domain (this_obj), field, this_obj, error);
-			if (!is_ok (error)) {
-				mono_error_set_pending_exception (error);
-				return 0;
-			}
-			values [count++] = o;
+				values = g_newa (MonoObjectHandle, mono_class_num_fields (klass));
```
Comment (lambdageek) [inline mono/metadata/icall.c]: The number of allocated handles is equal to the number of fields in a user-provided class. We try not to do that.

It's not clear to me why we need to get the fields ahead of time.  Can't we just use this first loop to count and assign the fields to the array in the second loop, below, without an intermediate stack allocated array?

File: mono/metadata/icall.c [hunk 3]
```diff
@@ -1530,19 +1524,13 @@ ves_icall_System_ValueType_Equals (MonoObject *this_obj, MonoObject *that, MonoA
...
-			}
-			values [count++] = o;
-			o = mono_field_get_value_object_checked (mono_object_domain (this_obj), field, that, error);
-			if (!is_ok (error)) {
-				mono_error_set_pending_exception (error);
-				return FALSE;
-			}
-			values [count++] = o;
+				values = g_newa (MonoObjectHandle, mono_class_num_fields (klass) * 2);
+			o = mono_field_get_value_object_checked (mono_handle_domain (this_obj), field, MONO_HANDLE_RAW (this_obj), error);
```
Comment (lambdageek) [inline mono/metadata/icall.c]: Same as `ves_icall_System_ValueType_InternalGetHashCode` - we don't want to allocate coop handles in proportion to a user-provided input.

Again, it's unclear to me that we really need to do the assignment in the first loop.

File: mono/metadata/sre.c [hunk 4]
```diff
@@ -4596,9 +4592,11 @@ ves_icall_CustomAttributeBuilder_GetBlob (MonoReflectionAssembly *assembly, Mono
 #endif
 
 void
-ves_icall_AssemblyBuilder_basic_init (MonoReflectionAssemblyBuilder *assemblyb)
+ves_icall_AssemblyBuilder_basic_init (MonoReflectionAssemblyBuilderHandle assemblyb, MonoError *error)
 {
-	mono_reflection_dynimage_basic_init (assemblyb);
+	uint32_t gchandle = mono_gchandle_from_handle (MONO_HANDLE_CAST (MonoObject, assemblyb), TRUE);
+	mono_reflection_dynimage_basic_init (MONO_HANDLE_RAW (assemblyb), error);
```
Comment (lambdageek) [inline mono/metadata/sre.c]: Expedient, but somewhat unsatisfying. :-D

But ok, fair enough.

**Comment by vargaz** [comment]
@monojenkins build failed

**Pull request merged**